loki/transform/transform_inline.py

```diff
@@ -238,17 +248,38 @@ def _map_unbound_dims(var, val):
                 if isinstance(arg, sym.Array):
                     # Resolve implicit dimension ranges of the passed value,
                     # eg. when passing a two-dimensional array `a` as `call(arg=a)`
-                    qualified_value = val if val.dimensions else val.clone(
-                        dimensions=tuple(sym.Range((None, None)) for _ in arg.shape)
-                    )
+                    # Check if val is a DeferredTypeSymbol, as it does not have a `dimensions` attribute
+                    if not isinstance(val, sym.DeferredTypeSymbol) and val.dimensions:
+                        qualified_value = val
+                    else:
+                        qualified_value = val.clone(
+                            dimensions=tuple(sym.Range((None, None)) for _ in arg.shape)
+                        )
                     arg_vars = tuple(v for v in member_vars if v.name == arg.name)
-                    argmap.update((v, _map_unbound_dims(v, qualified_value)) for v in arg_vars)
+                    argmap.update((v, val.clone(dimensions = _map_unbound_dims(v, qualified_value))) for v in arg_vars)
                 else:
                     argmap[arg] = val
 
             # Substitute argument calls into a copy of the body
             member_body = SubstituteExpressions(argmap).visit(member.body.body)
 
+            # Recurse for nested variables replacements
+            test = True
+            arg_names = [a.name for a in argmap]
+            depth = 0
+            while test :
+                depth += 1
+                test = False
+                for var in FindVariables().visit(member_body):
+                    if var.name in arg_names:
+                        test=True
+                if test:
+                    member_body = SubstituteExpressions(argmap).visit(member_body)
```

Comment:
I believe the `SubstituteExpression` utility already has option to apply recursion to the map itself. A depth-tested while loop is for sure an anti-pattern that I would very much like to avoid here (sorry).

The utility is here:
https://github.com/ecmwf-ifs/loki/blob/2322c1bd863b5f3e43157dde66f8c3f938a98680/loki/transform/transform_utilities.py#L417